# Battle of the Neighborhoods
## Coursera Capstone Project

In [1]:
import pandas as pd
import numpy as np
import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

## We make use of the data available on Wikipedia. 
Here we are using the link to the Wikipedia page hat has information about the various postal codes, Boroughs, and neighborhoods in Toronto. We will import this data into the Jupyter Notebook.

In [2]:
import requests
import urllib.request
import time
from bs4 import BeautifulSoup
import numpy as np
import pandas as pd
from urllib.request import urlopen
url = 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
page = urllib.request.urlopen(url)
soup = BeautifulSoup(page, "lxml")

## We are trying to import the Table from the Wikipedia page using Beautiful Soup

In [3]:
# use the 'find_all' function to bring back all instances of the 'table' tag in the HTML and store in 'all_tables' variable
all_tables=soup.find_all("table")


We know by inspecting the Wikipedia page that our desired table has <b>class = "wikitable sortable"</b><br>
We can use this information to select our required table and store it in right_table.

In [4]:
right_table=soup.find('table', class_='wikitable sortable')


## Converting into Pandas dataframe: 
Once we have our table stored in <b>right_table</b>, we can extract the columns and cell values from this table<br>
We know that each row begins with a <b>tr</b> tab and each cell value will begin with a <b>td</b> tab.<br>
using this information we can go through the table and extract the values one by one and store them into 3 columns which we can later rename when we add them into our pandas dataframe.

In [5]:
A=[]
B=[]
C=[]

for row in right_table.findAll('tr'):
    cells=row.findAll('td')
    if len(cells)==3:
        A.append(cells[0].find(text=True))
        B.append(cells[1].find(text=True))
        C.append(cells[2].find(text=True))

We now rename our columns and add them to our pandas data frame names <b>df</b><br>
The three columns are names <b>Postal Code</b>, <b>Borough</b> and <b>Neighborhood</b>

In [6]:
df=pd.DataFrame(A,columns=['Postal Code'])
df['Borough']=B
df['Neighborhood']=C
df.head(10)

Postal Code            Borough                                  Neighborhood
0        M1A
      Not assigned
                                 Not assigned

1        M2A
      Not assigned
                                 Not assigned

2        M3A
        North York
                                    Parkwoods

3        M4A
        North York
                             Victoria Village

4        M5A
  Downtown Toronto
                    Regent Park, Harbourfront

5        M6A
        North York
             Lawrence Manor, Lawrence Heights

6        M7A
  Downtown Toronto
  Queen's Park, Ontario Provincial Government

7        M8A
      Not assigned
                                 Not assigned

8        M9A
         Etobicoke
      Islington Avenue, Humber Valley Village

9        M1B
       Scarborough
                               Malvern, Rouge

#  Cleaning the data
We have to process the data so that we have the data  frame in the required  format

## We only process the cells that have an assigned borough. Ignore cells with a borough that is <b>Not assigned</b>.

In [7]:
df = df[df["Borough"]!='Not assigned\n']
df.head()

Postal Code            Borough                                  Neighborhood
2        M3A
        North York
                                    Parkwoods

3        M4A
        North York
                             Victoria Village

4        M5A
  Downtown Toronto
                    Regent Park, Harbourfront

5        M6A
        North York
             Lawrence Manor, Lawrence Heights

6        M7A
  Downtown Toronto
  Queen's Park, Ontario Provincial Government

## More than one neighborhood can exist in one postal code area. For example, in the table on the Wikipedia page, you will notice that M5A is listed twice and has two neighborhoods: Harbourfront and Regent Park. These two rows will be combined into one row with the neighborhoods separated with a comma.

In [8]:
df_combined = df.groupby("Postal Code", sort=False).agg( ','.join)
df_combined.head()

Borough                                   Neighborhood
Postal Code                                                                   
M3A                North York\n                                    Parkwoods\n
                   North York\n                             Victoria Village\n
M4A          Downtown Toronto\n                    Regent Park, Harbourfront\n
                   North York\n             Lawrence Manor, Lawrence Heights\n
M5A          Downtown Toronto\n  Queen's Park, Ontario Provincial Government\n
                                                                              
M6A                                                                           
                                                                              
M7A

We have to remove the '\n' character from the end of the Borough and Neighborhood column.<br>
We do this by splitting the data and reassigning the column values accordingly.

In [9]:
B = pd.DataFrame(df_combined.Borough.str.split('\n').str[0])
N = pd.DataFrame(df_combined.Neighborhood.str.split('\n').str[0])
df_combined['Borough'] = B["Borough"]
df_combined['Neighborhood'] = N["Neighborhood"]
df_combined.reset_index(inplace = True)
df_combined.head()

Postal Code           Borough                                 Neighborhood
0        M3A
        North York                                    Parkwoods
1        M4A
        North York                             Victoria Village
2        M5A
  Downtown Toronto                    Regent Park, Harbourfront
3        M6A
        North York             Lawrence Manor, Lawrence Heights
4        M7A
  Downtown Toronto  Queen's Park, Ontario Provincial Government

## If a cell has a borough but a Not assigned neighborhood, then the neighborhood will be the same as the borough.

In [10]:
df[df["Neighborhood"]=='Not assigned']

Empty DataFrame
Columns: [Postal Code, Borough, Neighborhood]
Index: []

From the above result, it is clear that there is no such neighborhood that is not assigned, but has a Borough. 

## We check the shape of the dataframe

In [11]:
df_combined.shape

(103, 3)

<b>The cleaned Data Set has 103 rows and 3 columns. </b>

We Sort the data based on Postal Code so that we can combine this data with its Latitudes and Longitudes

In [12]:
df_combined.sort_values(by = ["Postal Code"],inplace = True)

In [13]:
df_combined.reset_index(inplace=True)
df_combined.head(5)

index Postal Code      Borough                            Neighborhood
0      6        M1B
  Scarborough                          Malvern, Rouge
1     12        M1C
  Scarborough  Rouge Hill, Port Union, Highland Creek
2     18        M1E
  Scarborough       Guildwood, Morningside, West Hill
3     22        M1G
  Scarborough                                  Woburn
4     26        M1H
  Scarborough                               Cedarbrae

In [14]:
# The code was removed by Watson Studio for sharing.

## Adding the Latitudes and Longitudes
We load the data into Location data frame and then combine it with our data to complete our data frame

In [15]:
Location.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [16]:
df_combined["Latitude"] = Location["Latitude"]
df_combined["Longitude"] = Location["Longitude"]
df_combined.head()

index Postal Code      Borough                            Neighborhood  \
0      6        M1B
  Scarborough                          Malvern, Rouge   
1     12        M1C
  Scarborough  Rouge Hill, Port Union, Highland Creek   
2     18        M1E
  Scarborough       Guildwood, Morningside, West Hill   
3     22        M1G
  Scarborough                                  Woburn   
4     26        M1H
  Scarborough                               Cedarbrae   

    Latitude  Longitude  
0  43.806686 -79.194353  
1  43.784535 -79.160497  
2  43.763573 -79.188711  
3  43.770992 -79.216917  
4  43.773136 -79.239476

In [17]:
df_combined.reset_index(inplace=True)

In [18]:
df_combined.head()

level_0  index Postal Code      Borough  \
0        0      6        M1B
  Scarborough   
1        1     12        M1C
  Scarborough   
2        2     18        M1E
  Scarborough   
3        3     22        M1G
  Scarborough   
4        4     26        M1H
  Scarborough   

                             Neighborhood   Latitude  Longitude  
0                          Malvern, Rouge  43.806686 -79.194353  
1  Rouge Hill, Port Union, Highland Creek  43.784535 -79.160497  
2       Guildwood, Morningside, West Hill  43.763573 -79.188711  
3                                  Woburn  43.770992 -79.216917  
4                               Cedarbrae  43.773136 -79.239476

In [19]:
data = df_combined[["Postal Code","Borough","Neighborhood","Latitude","Longitude"]]
data.head()

Postal Code      Borough                            Neighborhood   Latitude  \
0        M1B
  Scarborough                          Malvern, Rouge  43.806686   
1        M1C
  Scarborough  Rouge Hill, Port Union, Highland Creek  43.784535   
2        M1E
  Scarborough       Guildwood, Morningside, West Hill  43.763573   
3        M1G
  Scarborough                                  Woburn  43.770992   
4        M1H
  Scarborough                               Cedarbrae  43.773136   

   Longitude  
0 -79.194353  
1 -79.160497  
2 -79.188711  
3 -79.216917  
4 -79.239476

This is our cleaned data frame ready to use for analysis.

## We want to  focus on the Neighborhoods that are in Toronto
We check to see what Boroughs have the word "Toronto" 

In [20]:
set(data["Borough"])

{'Central Toronto',
 'Downtown Toronto',
 'East Toronto',
 'East York',
 'Etobicoke',
 'Mississauga',
 'North York',
 'Scarborough',
 'West Toronto',
 'York'}

In [21]:
df_tor = data[data["Borough"].isin(["Central Toronto","Downtown Toronto","East Toronto","West Toronto"])]
df_tor.head()

Postal Code          Borough                    Neighborhood   Latitude  \
37        M4E
     East Toronto                     The Beaches  43.676357   
41        M4K
     East Toronto    The Danforth West, Riverdale  43.679557   
42        M4L
     East Toronto  India Bazaar, The Beaches West  43.668999   
43        M4M
     East Toronto                 Studio District  43.659526   
44        M4N
  Central Toronto                   Lawrence Park  43.728020   

    Longitude  
37 -79.293031  
41 -79.352188  
42 -79.315572  
43 -79.340923  
44 -79.388790

In [22]:
df_tor.reset_index(inplace = True)
df_tor = df_tor[["Postal Code","Borough","Neighborhood","Latitude","Longitude"]]
df_tor.head()

Postal Code          Borough                    Neighborhood   Latitude  \
0        M4E
     East Toronto                     The Beaches  43.676357   
1        M4K
     East Toronto    The Danforth West, Riverdale  43.679557   
2        M4L
     East Toronto  India Bazaar, The Beaches West  43.668999   
3        M4M
     East Toronto                 Studio District  43.659526   
4        M4N
  Central Toronto                   Lawrence Park  43.728020   

   Longitude  
0 -79.293031  
1 -79.352188  
2 -79.315572  
3 -79.340923  
4 -79.388790

##  Downloading the required packages to help us plot maps

In [23]:
!pip -q install folium
import folium 

In [24]:
latitude = 43.6532
longitude = -79.3832
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.6532, -79.3832.


## Create a map of Toronto with neighborhoods superimposed on top.

In [25]:
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, borough, neighborhood in zip(data['Latitude'], data['Longitude'], data['Borough'], data['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

<b>Folium</b> is a great visualization library. Feel free to zoom into the above map, and click on each circle mark to reveal the name of the neighborhood and its respective borough. <br>

However, for illustration purposes, let's simplify the above map and segment and cluster only the neighborhoods that have Borough names including Toronto. 

In [26]:
map_tor = folium.Map(location=[latitude, longitude], zoom_start=12)

# add markers to map
for lat, lng, borough, neighborhood in zip(df_tor['Latitude'], df_tor['Longitude'], df_tor['Borough'], df_tor['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_tor)  
    
map_tor

## Defining Foursquare Credentials

In [27]:
# The code was removed by Watson Studio for sharing.

### Let's explore the first neighborhood in our dataframe.
Get the neighborhood's name.

In [28]:
df_tor.loc[0, 'Neighborhood']

'The Beaches'

Get the neighborhood's latitude and longitude values.

In [29]:
neighborhood_latitude = df_tor.loc[0, 'Latitude'] # neighborhood latitude value
neighborhood_longitude = df_tor.loc[0, 'Longitude'] # neighborhood longitude value

neighborhood_name = df_tor.loc[0, 'Neighborhood'] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neighborhood_name, 
                                                               neighborhood_latitude, 
                                                               neighborhood_longitude))

Latitude and longitude values of The Beaches are 43.67635739999999, -79.2930312.


## Now, let's get the top 100 venues that are in The Beaches within a radius of 700 metres.
First, let's create the GET request URL.

In [30]:
LIMIT = 100 # limit of number of venues returned by Foursquare API
radius = 700 # define radius
# create URL
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighborhood_latitude, 
    neighborhood_longitude, 
    radius, 
    LIMIT)
url

'https://api.foursquare.com/v2/venues/explore?&client_id=SQXFWYKRR2LSWVB15F10GSC5L4D1ZGCFXQOF3OOFR54KWKME&client_secret=FKO5YKMB5DHQAE51YF2RJRDO2QQQGQ1STYQMWNDOSTDDJ3IV&v=20180605&ll=43.67635739999999,-79.2930312&radius=700&limit=100'

Send the GET request and examine the resutls

In [31]:
results = requests.get(url).json()

In [32]:
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

Now we are ready to clean the json and structure it into a pandas dataframe.

In [33]:
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

name                     categories        lat        lng
0  Glen Manor Ravine                          Trail  43.676821 -79.293942
1  Beaches Bake Shop                         Bakery  43.680363 -79.289692
2     The Beech Tree                      Gastropub  43.680493 -79.288846
3    Tori's Bakeshop  Vegetarian / Vegan Restaurant  43.672114 -79.290331
4    The Fox Theatre            Indie Movie Theater  43.672801 -79.287272

We check to see how many venues are returned by Foursquare : 

In [34]:
print('{} venues were returned by Foursquare.'.format(nearby_venues.shape[0]))

44 venues were returned by Foursquare.


## Explore Neighborhoods in Toronto

In [35]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

toronto_venues = getNearbyVenues(names=df_tor['Neighborhood'],
                                   latitudes=df_tor['Latitude'],
                                   longitudes=df_tor['Longitude']
                                  )

The Beaches
The Danforth West, Riverdale
India Bazaar, The Beaches West
Studio District
Lawrence Park
Davisville North
North Toronto West,  Lawrence Park
Davisville
Moore Park, Summerhill East
Summerhill West, Rathnelly, South Hill, Forest Hill SE, Deer Park
Rosedale
St. James Town, Cabbagetown
Church and Wellesley
Regent Park, Harbourfront
Garden District, Ryerson
St. James Town
Berczy Park
Central Bay Street
Richmond, Adelaide, King
Harbourfront East, Union Station, Toronto Islands
Toronto Dominion Centre, Design Exchange
Commerce Court, Victoria Hotel
Roselawn
Forest Hill North & West, Forest Hill Road Park
The Annex, North Midtown, Yorkville
University of Toronto, Harbord
Kensington Market, Chinatown, Grange Park
CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport
Stn A PO Boxes
First Canadian Place, Underground city
Christie
Dufferin, Dovercourt Village
Little Portugal, Trinity
Brockton, Parkdale Village, Exhibition Place
High

In [36]:
print(toronto_venues.shape)
toronto_venues.head()

(1640, 7)


Neighborhood  Neighborhood Latitude  \
0                   The Beaches              43.676357   
1                   The Beaches              43.676357   
2                   The Beaches              43.676357   
3                   The Beaches              43.676357   
4  The Danforth West, Riverdale              43.679557   

   Neighborhood Longitude                               Venue  Venue Latitude  \
0              -79.293031                   Glen Manor Ravine       43.676821   
1              -79.293031  The Big Carrot Natural Food Market       43.678879   
2              -79.293031                 Grover Pub and Grub       43.679181   
3              -79.293031                       Upper Beaches       43.680563   
4              -79.352188                       MenEssentials       43.677820   

   Venue Longitude     Venue Category  
0       -79.293942              Trail  
1       -79.297734  Health Food Store  
2       -79.297215                Pub  
3       -79.292869       Neighborhood  
4       -79.351265     Cosmetics Shop

In [37]:
toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Berczy Park                                                            60   
Brockton, Parkdale Village, Exhibition Place                           26   
Business reply mail Processing Centre, South Ce...                     18   
CN Tower, King and Spadina, Railway Lands, Harb...                     17   
Central Bay Street                                                     64   
Christie                                                               17   
Church and Wellesley                                                   77   
Commerce Court, Victoria Hotel                                        100   
Davisville                                                             34   
Davisville North                                                       10   
Dufferin, Dovercourt Village                                           15   
First Canadian Place, Underground city                                100   
Forest Hill North & West, Forest Hill Road Park                         4   
Garden District, Ryerson                                              100   
Harbourfront East, Union Station, Toronto Islands                     100   
High Park, The Junction South                                          24   
India Bazaar, The Beaches West                                         20   
Kensington Market, Chinatown, Grange Park                              65   
Lawrence Park                                                           4   
Little Portugal, Trinity                                               46   
Moore Park, Summerhill East                                             4   
North Toronto West,  Lawrence Park                                     17   
Parkdale, Roncesvalles                                                 14   
Queen's Park, Ontario Provincial Government                            34   
Regent Park, Harbourfront                                              45   
Richmond, Adelaide, King                                              100   
Rosedale                                                                4   
Roselawn                                                                2   
Runnymede, Swansea                                                     39   
St. James Town                                                         87   
St. James Town, Cabbagetown                                            40   
Stn A PO Boxes                                                         97   
Studio District                                                        40   
Summerhill West, Rathnelly, South Hill, Forest ...                     17   
The Annex, North Midtown, Yorkville                                    20   
The Beaches                                                             4   
The Danforth West, Riverdale                                           41   
Toronto Dominion Centre, Design Exchange                              100   
University of Toronto, Harbord                                         34   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Berczy Park                                                             60   
Brockton, Parkdale Village, Exhibition Place                            26   
Business reply mail Processing Centre, South Ce...                      18   
CN Tower, King and Spadina, Railway Lands, Harb...                      17   
Central Bay Street                                                      64   
Christie                                                                17   
Church and Wellesley                                                    77   
Commerce Court, Victoria Hotel                                         100   
Davisville                                                              34   
Davisville North                    

Let's find out how many unique categories can be curated from all the returned venues

In [38]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 238 uniques categories.


## Next, let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category

In [39]:
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")
# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 
#move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped.shape

(39, 238)

Let's print each neighborhood along with the top 5 most common venues

In [40]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                        Berczy Park           Coffee Shop   
1       Brockton, Parkdale Village, Exhibition Place                  Café   
2  Business reply mail Processing Centre, South C...    Light Rail Station   
3  CN Tower, King and Spadina, Railway Lands, Har...       Airport Service   
4                                 Central Bay Street           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue  4th Most Common Venue  \
0                  Café        Farmers Market     Seafood Restaurant   
1                Bakery        Breakfast Spot  Performing Arts Venue   
2               Brewery      Recording Studio             Restaurant   
3        Airport Lounge           Coffee Shop        Harbor / Marina   
4        Sandwich Place                  Café     Italian Restaurant   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0                Bakery          Cocktail Bar              Beer Bar   
1           Coffee Shop                   Gym  Gym / Fitness Center   
2  Fast Food Restaurant        Farmers Market         Auto Workshop   
3              Boutique         Boat or Ferry   Rental Car Location   
4   Japanese Restaurant      Department Store          Burger Joint   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0           Cheese Shop            Restaurant     Italian Restaurant  
1             Pet Store             Nightclub     Italian Restaurant  
2               Butcher                   Spa          Burrito Place  
3                   Bar                 Plane       Sculpture Garden  
4       Bubble Tea Shop           Salad Place  Portuguese Restaurant

## Cluster Neighborhoods
Run k-means to cluster the neighborhood into 5 clusters.

In [41]:
# import k-means from clustering stage
from sklearn.cluster import KMeans
import matplotlib.cm as cm
import matplotlib.colors as colors

# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = df_tor

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

toronto_merged.head() # check the last columns!

Postal Code          Borough                    Neighborhood   Latitude  \
0        M4E
     East Toronto                     The Beaches  43.676357   
1        M4K
     East Toronto    The Danforth West, Riverdale  43.679557   
2        M4L
     East Toronto  India Bazaar, The Beaches West  43.668999   
3        M4M
     East Toronto                 Studio District  43.659526   
4        M4N
  Central Toronto                   Lawrence Park  43.728020   

   Longitude  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0 -79.293031               0                   Pub                 Trail   
1 -79.352188               2      Greek Restaurant           Coffee Shop   
2 -79.315572               0        Sandwich Place           Pizza Place   
3 -79.340923               2                  Café           Coffee Shop   
4 -79.388790               0                  Park    Dim Sum Restaurant   

  3rd Most Common Venue   4th Most Common Venue 5th Most Common Venue  \
0     Health Food Store           Women's Store        Discount Store   
1    Italian Restaurant  Furniture / Home Store            Restaurant   
2                   Gym       Food & Drink Shop               Brewery   
3               Brewery               Gastropub                Bakery   
4              Bus Line             Swim School         Women's Store   

  6th Most Common Venue        7th Most Common Venue 8th Most Common Venue  \
0      Department Store                 Dessert Shop    Dim Sum Restaurant   
1        Ice Cream Shop               Cosmetics Shop               Brewery   
2         Burrito Place           Italian Restaurant            Restaurant   
3   American Restaurant            Convenience Store        Sandwich Place   
4      Department Store  Eastern European Restaurant   Dumpling Restaurant   

  9th Most Common Venue 10th Most Common Venue  
0                 Diner                Dog Run  
1       Bubble Tea Shop                   Café  
2  Fast Food Restaurant             Steakhouse  
3           Cheese Shop         Clothing Store  
4            Donut Shop       Doner Restaurant

In [42]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

We can see that most neighborhoods end up falling in the same cluster. So there is not much difference between the places in Toronto when it comes to what places are popular among the tourists and local people.